timemachine/fe/chiral_utils.py

```diff
@@ -132,6 +147,136 @@ def find_chiral_atoms(mol):
     return chiral_atoms
 
 
+def setup_all_chiral_atom_restr_idxs(mol, conf) -> List[FourTuple]:
+    """Apply setup_chiral_atom_restraints to all atoms found by find_chiral_atoms"""
+    chiral_atom_set = find_chiral_atoms(mol)
+    chiral_atom_restr_idxs = []
+    for a_idx in chiral_atom_set:
+        idxs = setup_chiral_atom_restraints(mol, conf, a_idx)
+        for ii in idxs:
+            assert ii not in chiral_atom_restr_idxs
+        chiral_atom_restr_idxs.extend(idxs)
+    return chiral_atom_restr_idxs
+
+
+class ChiralRestrIdxSet:
+    """Support fast checks of whether a trial 4-tuple is consistent with a set of chiral atom idxs"""
+
+    def __init__(self, restr_idxs: List[FourTuple]):
+        self.restr_idxs = restr_idxs
+        self.allowed_set, self.disallowed_set = self.expand_symmetries()
+
+    @classmethod
+    def from_mol(cls, mol, conf):
+        restr_idxs = setup_all_chiral_atom_restr_idxs(mol, conf)
+        return ChiralRestrIdxSet(restr_idxs)
+
+    def expand_symmetries(self) -> Tuple[Set[FourTuple], Set[FourTuple]]:
+        allowed_set = set()
+        disallowed_set = set()
+
+        for (center, i, j, k) in self.restr_idxs:
+            # rotations
+            allowed_set.add((center, i, j, k))
+            allowed_set.add((center, j, k, i))
+            allowed_set.add((center, k, i, j))
+
+            # swaps
+            disallowed_set.add((center, i, k, j))
+            disallowed_set.add((center, j, i, k))
+            disallowed_set.add((center, k, j, i))
+
+        assert allowed_set.isdisjoint(disallowed_set)
+
+        return allowed_set, disallowed_set
+
+    def defines(self, trial_tuple: FourTuple) -> bool:
+        return (trial_tuple in self.allowed_set) or (trial_tuple in self.disallowed_set)
```

Comment:
minor nit: would it be slightly more efficient to represent the allowed and disallowed sets as a `is_allowed: Dict[FourTuple, bool]`? (e.g. `idxs in allowed_set or idxs in disallowed_set === idxs in is_allowed` and `idxs in disallowed_set === not is_allowed.get(idxs)`

Reply:
Hmm, haven't profiled for efficiency differences like that yet.

Sets seem marginally simpler to read than dicts (the only important thing added by this PR is [`trial_tuple in self.disallowed_set`](https://github.com/proteneer/timemachine/pull/908/files/7787f1e2a9036da81ec363b9ab5a524912a568f0#diff-cc54470389f7198a698e6ff9a5a6d7886cf32a4dc9c96bdd7a6c2f25d161e421R197)), but either spelling would be easily readable imo.

I would have a slight preference to leave this as a set for now, and defer performance to a second (profile-guided) pass, unless there's a readability advantage to switching now.

Reply:
Sounds good!